### ISSUE TITLE
test MMS/spatial/advection is broken

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#1503]
Should be fixed or disabled/removed

https://omadesala.alfa1024.info/BEST/2019-01-15_20:01/

**Comment by dschwoerer** [comment]
Cvode doesn't converge:
```
advection.cxx: In member function ‘virtual int AdvectMMS::init(bool)’:
advection.cxx:12:17: warning: unused parameter ‘restarting’ [-Wunused-parameter]
   int init(bool restarting) {
            ~~~~~^~~~~~~~~~
rm: cannot remove 'data/BOUT.dmp.*.nc': No such file or directory
/usr/lib64/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
getmpirun: using the default mpirun -np
Making MMS steady-state advection test
  Running with method=2 mesh:nx=20 mesh:dx=0.39269908169872414 MZ=16
  -> Error norm: l-2 0.383543 l-inf 0.983098
  Running with method=2 mesh:nx=36 mesh:dx=0.19634954084936207 MZ=32
  -> Error norm: l-2 0.046825 l-inf 0.133331
Convergence order = 3.034030
... Failure
  Running with method=0 mesh:ddx:upwind=u1 mesh:ddz:upwind=u1 mesh:nx=20 mesh:dx=0.39269908169872414 MZ=16
  -> Error norm: l-2 0.408177 l-inf 1.012092
  Running with method=0 mesh:ddx:upwind=u1 mesh:ddz:upwind=u1 mesh:nx=36 mesh:dx=0.19634954084936207 MZ=32
  -> Error norm: l-2 0.096668 l-inf 0.253686
Convergence order = 2.078086
... Failure
  Running with method=0 mesh:ddx:upwind=c2 mesh:ddz:upwind=c2 mesh:nx=20 mesh:dx=0.39269908169872414 MZ=16
  -> Error norm: l-2 0.388361 l-inf 0.998959
  Running with method=0 mesh:ddx:upwind=c2 mesh:ddz:upwind=c2 mesh:nx=36 mesh:dx=0.19634954084936207 MZ=32
  -> Error norm: l-2 0.047559 l-inf 0.131730
Convergence order = 3.029612
... Failure
  Running with method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=20 mesh:dx=0.39269908169872414 MZ=16
Traceback (most recent call last):
  File "./runtest", line 68, in <module>
    s, out = launch_safe(cmd, runcmd=MPIRUN, nproc=nproc, pipe=True)
  File "/home/bout-test/tmp.i6YN366nkJ/tools/pylib/boututils/run_wrapper.py", line 275, in launch_safe
    (s,command,out))
RuntimeError: Run failed with 1.
Command was:
./advection method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=20 mesh:dx=0.39269908169872414 MZ=16

Output was

LOCALE_PATH = '/usr/local/share/locale'
LOCALE_PATH = '/usr/local/share/locale'
BOUT++ version 4.2.0
Revision: 6caa408738701cd187ae7bb862622769d8727feb
MD5 checksum: 61572ef1e80006aab26e768ce6826f01
Code compiled on Jan 15 2019 at 20:03:34

B.Dudson (University of York), M.Umansky (LLNL) 2007
Based on BOUT by Xueqiao Xu, 1999

Processor number: 0 of 2

pid: 12875

Compile-time options:
	Checking enabled, level 2
	Signal handling enabled
	netCDF4 support enabled
	Parallel NetCDF support disabled
	OpenMP parallelisation disabled
	Compiled with flags : "-DMD5SUM=61572ef1e80006aab26e768ce6826f01 -DREVISION=6caa408738701cd187ae7bb862622769d8727feb -g -Wall -Wextra -Wnull-dereference -Wno-cast-function-type -O2 -DCHECK=2 -DSIGHANDLE -DLOGCOLOR -DBOUT_VERSION_STRING=\"4.2.0\" -DBOUT_VERSION_DOUBLE=4.20 -DBACKTRACE -fPIC -DBOUT_HAS_FFTW -DNCDF4 -DHDF5 -DBOUT_HAS_PVODE -DBOUT_HAS_GETTEXT -DBOUT_LOCALE_PATH=/usr/local/share/locale"
	Command line options for this run : ./advection method=0 mesh:ddx:upwind=w3 mesh:ddz:upwind=w3 mesh:nx=20 mesh:dx=0.39269908169872414 MZ=16 
Reading options file data/BOUT.inp
	Option mesh:ddx:upwind = c2 (data/BOUT.inp) overwritten with:
		mesh:ddx:upwind = w3 (Command line)
	Option mesh:ddz:upwind = c2 (data/BOUT.inp) overwritten with:
		mesh:ddz:upwind = w3 (Command line)
	Option mesh:nx = 18 (data/BOUT.inp) overwritten with:
		mesh:nx = 20 (Command line)
	Option mesh:dx = 0.3926990816 (data/BOUT.inp) overwritten with:
		mesh:dx = 0.39269908169872414 (Command line)
Writing options to file data/BOUT.settings

Getting grid data from options
	Option mesh:type = bout (default)
	Option mesh:StaggerGrids = 0 (default)
	Option mesh:maxregionblocksize = 64 (default)
	Option mesh:ddx:upwind = w3 (Command line)
	Option mesh:ddz:upwind = w3 (Command line)
	Option mesh:ddx:upwind = w3 (Command line)
	Option mesh:ddz:upwind = w3 (Command line)
	Option mesh:ddz:fft_filter = 0 (default)
	Option mesh:symmetricglobalx = true (data/BOUT.inp)
	Option optionfile = BOUT.inp
WARNING: The default of this option has changed in release 4.1.
If you want the old setting, you have to specify mesh:symmetricGlobalY=false in BOUT.inp
	Option mesh:symmetricGlobalY = 1 (default)
Loading mesh
	Option mesh:nx = 20 (Command line)
	Option mesh:ny = 1 (data/BOUT.inp)
	Option mz = 16 (Command line)
	Grid size: 20 x 1 x 16
	Option mxg = 2 (data/BOUT.inp)
	Option myg = 1 (data/BOUT.inp)
	Guard cells (x,y): 2, 1
	WARNING: Separatrix location 'ixseps1' not found. Setting to 20
	WARNING: Separatrix location 'ixseps2' not found. Setting to 20
	WARNING: Branch-cut 'jyseps1_1' not found. Setting to -1
	WARNING: Branch-cut 'jyseps1_2' not found. Setting to 0
	WARNING: Branch-cut 'jyseps2_1' not found. Setting to 0
	WARNING: Branch-cut 'jyseps2_2' not found. Setting to 0
	WARNING: Number of inner y points 'ny_inner' not found. Setting to 0
Finding value for NXPE (ideal = 5.656854)
	Candidate value: 2
	 -> Good value
	Domain split (NXPE=2, NYPE=1) into domains (localNx=8, localNy=1)
	Option IncIntShear = 0 (default)
	Option periodicX = 0 (default)
	Option async_send = 0 (default)
	Option ZMIN = 0 (default)
	Option zmax = 1 (data/BOUT.inp)
	EQUILIBRIUM IS SINGLE NULL (SND) 
Connection between top of Y processor 0 and bottom of 0 in range 0 <= x < 20
=> This processor sending in up
=> This processor sending in down
WARNING adding connection: poloidal index -1 out of range
	MYPE_IN_CORE = 1
	DXS = 12, DIN = 0. DOUT = -1
	UXS = 12, UIN = 0. UOUT = -1
	XIN = -1, XOUT = 1
	Twist-shift: DI UI 
	Option TwistShift = 0 (default)
Boundary regions in this processor: core, 
Constructing default regions
	Boundary region inner X
	Boundary region outer X
	done
	Option mesh:paralleltransform = identity (default)
	Option append = 0 (default)
	Option dump_format = nc (default)
Setting up output (dump) file
	Option output:parallel = 0 (default)
	Option output:flush = 1 (default)
	Option output:guards = 1 (default)
	Option output:floats = false (data/BOUT.inp)
	Option output:openclose = 1 (default)
	Option output:enabled = 1 (default)
	Option output:init_missing = 0 (default)
	Option output:shiftOutput = 0 (default)
	Option output:shiftInput = 0 (default)
	Option output:flushFrequency = 1 (default)
	Using NetCDF4 format for file 'data/BOUT.dmp.nc'
	Option mesh:dx = 0.39269908169872414 (Command line)
Variable 'dy' not in mesh options. Setting to 0.000000e+00
	WARNING: differencing quantity 'dy' not found. Set to 1.0
	Option ZMIN = 0 (default)
	Option zmax = 1 (data/BOUT.inp)
Variable 'g11' not in mesh options. Setting to 1.000000e+00
Variable 'g22' not in mesh options. Setting to 1.000000e+00
Variable 'g33' not in mesh options. Setting to 1.000000e+00
Variable 'g12' not in mesh options. Setting to 0.000000e+00
Variable 'g13' not in mesh options. Setting to 0.000000e+00
Variable 'g23' not in mesh options. Setting to 0.000000e+00
	Local maximum error in diagonal inversion is 0.000000e+00
	Local maximum error in off-diagonal inversion is 0.000000e+00
Variable 'J' not in mesh options. Setting to 0.000000e+00
	WARNING: Jacobian 'J' not found. Calculating from metric tensor
Variable 'Bxy' not in mesh options. Setting to 0.000000e+00
	WARNING: Magnitude of B field 'Bxy' not found. Calculating from metric tensor
Calculating differential geometry terms
	Communicating connection terms
	Option non_uniform = 1 (default)
Variable 'd2x' not in mesh options. Setting to 0.000000e+00
	WARNING: differencing quantity 'd2x' not found. Calculating from dx
Variable 'd2y' not in mesh options. Setting to 0.000000e+00
	WARNING: differencing quantity 'd2y' not found. Calculating from dy
Variable 'ShiftTorsion' not in mesh options. Setting to 0.000000e+00
	WARNING: No Torsion specified for zShift. Derivatives may not be correct
	Option restart:parallel = 0 (default)
	Option restart:flush = 1 (default)
	Option restart:guards = 1 (default)
	Option restart:floats = 0 (default)
	Option restart:openclose = 1 (default)
	Option restart:enabled = 1 (default)
	Option restart:init_missing = 0 (default)
	Option restart:shiftOutput = 0 (default)
	Option restart:shiftInput = 0 (default)
	Option restart:flushFrequency = 1 (default)
	Option solver:type =  (default)
	Option solver:monitor_timestep = 0 (default)
	Option solver:mms = true (data/BOUT.inp)
	Option solver:mms_initialise = false (data/BOUT.inp)
	Option restart = 0 (default)
Setting boundary for variable f
	core region: 	Option f:bndry_all = dirichlet_o2(f:solution) (data/BOUT.inp)
	Option f:solution = cos(4.0*x^2 + z) (data/BOUT.inp)

	Option f:function = 0 (data/BOUT.inp)
	Option f:scale = 0 (data/BOUT.inp)
	Option all:evolve_bndry = 0 (default)
	Option f:evolve_bndry = 0 (default)
	Option method = 0 (Command line)
	Option datadir = data
	Option dump_format = nc (default)
	Option restart_format = nc (default)
	Using NetCDF4 format for file 'data/BOUT.restart.nc'
	Option nout = 1 (data/BOUT.inp)
	Option timestep = 50 (data/BOUT.inp)
	Option solver:NOUT = 1 (default)
	Option solver:output_step = 50 (default)
Solver running for 1 outputs with output timestep of 5.000000e+01
Initialising solver
Initialising PVODE solver
	3d fields = 1, 2d fields = 0 neq=256, local_N=128
	Option solver:mudq = 10 (default)
	Option solver:mldq = 10 (default)
	Option solver:mukeep = 0 (default)
	Option solver:mlkeep = 0 (default)
	Option solver:atol = 1e-15 (data/BOUT.inp)
	Option solver:rtol = 1e-15 (data/BOUT.inp)
	Option solver:use_precon = 0 (default)
	Option solver:precon_dimens = 50 (default)
	Option solver:precon_tol = 0.0001 (default)
	Option solver:mxstep = 10000 (data/BOUT.inp)
	Option g:solution = sin(6.0*x^2 - z) (data/BOUT.inp)
	Option f:source = 3.18309886183791*x*sin(4.0*x^2 + z)*cos(6.0*x^2 - z) (data/BOUT.inp)
Running simulation


Run started at  : Tue 15 Jan 2019 20:27:22 GMT
	Option restart = false (default)
	Option append = false (default)
	Option dump_on_restart = 1 (default)

	Option wall_limit = -1 (default)
	Option stopCheck = 0 (default)
Sim Time  |  RHS evals  | Wall Time |  Calc    Inv   Comm    I/O   SOLVER

0.000e+00          2       1.57e-02     2.5    0.0    5.0   38.8   53.7
|  Step 0 of 1. Elapsed 0:00:00.0 ETA 0:00:00.0CVode-- At t=18.6445, mxstep=10000 steps taken on this call before
reaching tout=50.

CVode-- At t=18.6445, mxstep=10000 steps taken on this call before
reaching tout=50.

ERROR CVODE step failed, flag = -3
Timestep failed. Aborting
Error encountered in solver run
PVODE timestep failed
Error encountered
application called MPI_Abort(comm=0x84000002, 1) - process 1
```

**Comment by ZedThree** [comment]
Ignoring CVODE not converging, all the methods fail because they are one order higher than expected.

@d7919 and I have been investigating a bit, trying the method that fixed #1528 but that doesn't seem to be sufficient here.

It's quite sensitive to both the size of the dissipation term and the exact time solver and its options.

@bendudson How important is this steady-state test vs the time-dependent test in MMS/advection, and vs just computing the exact solution?

**Comment by bendudson** [comment]
I think the steady-state test was intended to make it (relatively) insensitive to the time integration method, so it's just testing the spatial operators. Just computing the exact solution should achieve the same aim. It sounds like this test is numerically ill conditioned, if it's very sensitive to dissipation terms.

**Comment by ZedThree** [comment]
I've replaced the test with just calculating the exact solution, and everything but WENO3 passes. Weirdly, W3 seems to have two regions for the error, neither of which is third order!

![advection_norm_l2](https://user-images.githubusercontent.com/1486942/52139606-8311e680-2648-11e9-987f-3faeef73cf9a.png)

Going up to 2048 points gives an order of 3.3 for the last two points:

![advection_norm_l2](https://user-images.githubusercontent.com/1486942/52139691-c10f0a80-2648-11e9-866a-71ae9f01a657.png)

**Comment by d7919** [comment]
> Going up to 2048 points gives an order of 3.3 for the last two points:

This is the same order that is seen for W3 across a large range of grid sizes in the `MMS/advection/weno3` test.

**Comment by ZedThree** [comment]
Fixed in #1540 (although W3 still has problems, this is probably due to it having a weird scaling)
